R/check_gene_expression.R

```diff
@@ -0,0 +1,111 @@
+      filter(!is.na(mrna_sample_id)) %>% 
+      with( split(mrna_sample_id, sample_seqType) ) %>% 
+      .[lengths(.) > 1] %>% 
+      lapply(unique) %>% 
+      .[lengths(.) > 1] %>% 
+      names %>% 
+      sub(" .+", "", .) %>% 
+      { mutate(expression_data_rows, multi_exp = ifelse(biopsy_id %in% ., 1, 0)) }
+  }
+  
+  expression_data_rows = mark_duplicates_biopsy(expression_data_rows)
+  
+  if(verbose){
+    now_row = nrow(expression_data_rows)
+    print(paste(now_row,"remain after collapsing on biopsy_id"))
+  }
+  
+  # collapse duplicates if any
+  if(any(expression_data_rows$multi_exp == 1) ){
+    if(verbose){
+      rows_total = filter(expression_data_rows,multi_exp==1) %>% nrow()
+      unique_redundant = filter(expression_data_rows,multi_exp==1) %>% pull(biopsy_id) %>% unique()
+      nid = length(unique_redundant)
+      print(paste(rows_total, "LINE 74: rows corresponding to ", nid, "unique biopsies are redundant, dropping redundancy"))
+      print(head(unique_redundant))
+    }
+    original = expression_data_rows %>% filter(!is.na(mrna_sample_id))
+    expression_data_rows = group_by(expression_data_rows, patient_id, biopsy_id) %>% 
+      slice_max(str_detect(protocol, "Ribo"), n=1, with_ties = TRUE) %>% 
+      slice_max(ffpe_or_frozen == "frozen", n=1, with_ties = TRUE) %>% 
+      ungroup()
+    
+    
+    expression_data_rows = mark_duplicates_biopsy(expression_data_rows)
+    
+    if(verbose){
+      now_row = nrow(expression_data_rows)
+      print(paste(now_row,"remain after collapsing on biopsy_id"))
+    }
+
+  }else{
+    lost = data.frame()
+  }
+  expression_data_rows = 
+    mutate(expression_data_rows,sample_id = mrna_sample_id) %>% 
+    unique()
+  if(!show_linkages){
+    return(select(expression_data_rows,-multi_exp))
+  }
+  #Optionally join to other seq_types where possible in case a user wants to see which samples have both RNA and a source of mutations
+  these_samples_metadata = get_gambl_metadata(seq_type_filter = c("genome","capture"))  %>%
+    dplyr::select(sample_id, 
+                  patient_id, biopsy_id, seq_type)
+  capture_meta = dplyr::filter(these_samples_metadata,
+                               seq_type=="capture",
+                               patient_id %in% expression_data_rows$patient_id) %>%
+    select(sample_id,patient_id,biopsy_id) %>%
+    rename(c("capture_sample_id"="sample_id"))
+  expression_data_rows = left_join(expression_data_rows,capture_meta,by=c("patient_id","biopsy_id"))
+  
+  genome_meta =  dplyr::filter(these_samples_metadata,
+                               seq_type=="genome",
+                               patient_id %in% expression_data_rows$patient_id) %>%
+    select(sample_id,patient_id,biopsy_id) %>%
+    rename(c("genome_sample_id"="sample_id"))
+  expression_data_rows = left_join(expression_data_rows,genome_meta, by=c("patient_id","biopsy_id"))
+  
+  return(select(expression_data_rows,-multi_exp))
+}
```

Comment:
The outputs below show that the `mark_duplicates_biopsy` function is accomplishing what it's supposed to (right?). 
The second output also shows that linking to the capture and genome sample doesn't cause any artificial duplication of rows. I've set this up as an optional output that is not returned by default. 
```
> check_gene_expression(verbose = T,show_linkages = F) %>% nrow()
[1] "starting with 2630 rows"                                                                                                                                                                               
[1] "2630 remain after collapsing on biopsy_id"
[1] "362 LINE 74: rows corresponding to  181 unique biopsies are redundant, dropping redundancy"
[1] "S01-4978"  "SU01-6860" "AS01-7426" "S01-12869" "S03-5298"  "US02-56"  
[1] "2449 remain after collapsing on biopsy_id"
[1] 2449
> check_gene_expression(verbose = T,show_linkages = T) %>% nrow()
[1] "starting with 2630 rows"                                                                                                                                                                               
[1] "2630 remain after collapsing on biopsy_id"
[1] "362 LINE 74: rows corresponding to  181 unique biopsies are redundant, dropping redundancy"
[1] "S01-4978"  "SU01-6860" "AS01-7426" "S01-12869" "S03-5298"  "US02-56"  
[1] "2449 remain after collapsing on biopsy_id"
[1] 2449                                
```

Reply:
I see that this is true, but now I am worried that `get_gambl_metadata` is not capable of returning metadata for some capture sample IDs...